v2/internal/util/match/glob_matcher.go

```diff
@@ -3,46 +3,42 @@
  * Licensed under the MIT license.
  */
 
-package config
+package match
 
 import (
-	"fmt"
 	"regexp"
 	"strings"
-	"sync"
 
-	"github.com/Azure/azure-service-operator/v2/internal/set"
 	"github.com/pkg/errors"
+
+	"github.com/Azure/azure-service-operator/v2/internal/set"
 )
 
 // A globMatcher is used to match a string against a literal or wildcard
 type globMatcher struct {
 	glob       string         // The glob we're matching, may contain wildcards * and ?
 	regex      *regexp.Regexp // A regular expression to match the glob
-	once       sync.Once
 	matched    bool
 	candidates set.Set[string]
 }
 
 var _ StringMatcher = &globMatcher{}
 
-// newGlobMatcher returns a new matcher for handling wildcards
-func newGlobMatcher(glob string) *globMatcher {
-	// Guard against misuse
-	if !HasWildCards(glob) {
-		msg := fmt.Sprintf("glob string %q has no wildcards", glob)
-		panic(msg)
+// NewGlobMatcher returns a new matcher for handling wildcards
+func NewGlobMatcher(glob string) (StringMatcher, error) {
+	regex, err := newGlobRegex(glob)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create regexp for glob %q", glob)
 	}
 
 	return &globMatcher{
 		glob:       glob,
+		regex:      regex,
 		candidates: make(set.Set[string]),
-	}
+	}, nil
```

Comment:
The panic was correct.

This factory method should remain private (`newGlobMatcher`), and only be called when the predicate `HasWildCards()` returns true. It's a programming bug if the glob string has no wildcards, not an error.

Reply:
r.e. private, I think that makes sense.

> It's a programming bug if the glob string has no wildcards, not an error.

"myString" is a valid glob. It's a glob with no wildcards but that's ok, it works just fine? How is it a programming bug?

Reply:
A wildcard matcher should only ever be used with a match string containing wildcards. 

If it's called with anything else, that's an error in programming. The way this _was_ ensured was that the factory method was only ever called from one place, and that call protected by the `HasWildcards()` predicate.

Reply:
Now that this is private again (I had made it public in an earlier refactor and then when I moved things into the new `match` package forgotten to make them private again), the above is true again.

I've brought the `HasWildCards` check back at the top of this constructor as well along with the panic.